### ISSUE TITLE
Using typst as a library, and passing file content to it directly in memory

### ISSUE THREAD

**Issue by rousbound** [typst/typst#1158]
### Description

In the context of integrating typst in systems for dynamic document generation, it would be useful to pass certain files to it directly, without needing to make the compiler read the files from disk.

### Use Case

I'm going to make an example in python, for the sake of simplicity and because there's[ typst-py](https://github.com/messense/typst-py)(Which might benefit from this also). But I also wish to do the equivalent of the following code  in rust, using typst as a library:

```python
import typst

compiler = typst.Compiler(".")
json_file = get_json_file() # generates a json file out of a database for example

external_data = typst.data_loader()
# Variable name
# Variable type (Can be json, txt, csv, toml, etc)
# Variable content
external_data.add("my_json_file", "json", json_file)

pdf_bytes = typst.compile("hello.typ", external_data)
``

So in the document I could retrieve the json file in some way like this:

#for value in $my_json_file { // maybe use '$' to express external variables?
...
} 

Don't know if this would require some deep change to the program architecture or the incremental compilation system though.

**Comment by PgBiel** [comment]
I'm pretty sure that's already possible, you just need to add it to the global scope when creating the compiler's World

**Comment by Myriad-Dreamin** [comment]
replace this function by reading your managed file content in memory

https://github.com/typst/typst/blob/b6a241cad232d920154986781c6f09d83dc1b6f6/cli/src/main.rs#L566-L574

**Comment by rousbound** [comment]
> replace this function by one reading your managed file content
> 
> https://github.com/typst/typst/blob/b6a241cad232d920154986781c6f09d83dc1b6f6/cli/src/main.rs#L566-L574

If I understood you correctly, you are saying that I could change this function so I can give as input the ".typ" file from memory.

 But that's not the problem, I actually want my template ".typ" file to be read from disk, only the json file  to be included in the "typ" file to be passed to typst from memory.

**Comment by rousbound** [comment]
> It is possible better (in typst style) to write a function that convert path.
> 
> ```
> // typst
> #let load(path, loader: "default") = { .. }
> #load("a.csv")
> ```

I'm sorry, I think I did not understand your solution. Can you provide a more concrete example?

**Comment by PgBiel** [comment]
> But that's not the problem, I actually want my template ".typ" file to be read from disk, only the json file I want to be passed to typst from memory.

yeah, you have to modify this line:
https://github.com/typst/typst/blob/b6a241cad232d920154986781c6f09d83dc1b6f6/cli/src/main.rs#L427

**Comment by rousbound** [comment]
I was going to reply to your first comment right now. It was very insightful. 

> > But that's not the problem, I actually want my template ".typ" file to be read from disk, only the json file I want to be passed to typst from memory.
> 
> yeah, you have to modify this line:
> 
> https://github.com/typst/typst/blob/b6a241cad232d920154986781c6f09d83dc1b6f6/cli/src/main.rs#L427



Maybe this solution works, if I can access the typst_library global module and create a new definition.

Then I would need to import into the typst/cli/main.rs some function from typst-library to convert my desired data into a typst value.

It is a bit hacky though.

Is this what you're thinking?

I will take a look at it.

**Comment by PgBiel** [comment]
> Is this what you're thinking?

Mostly, except that you probably don't need to modify the library itself, just take the output of the `build()` function, which will be a struct with a `.global` field which you can modify and add any variable you want there (it contains a `Scope` where you can use `.define("name", "something".into())`).

**Comment by rousbound** [comment]
> I was going to reply to your first comment right now. It was very insightful.
> 
> > > But that's not the problem, I actually want my template ".typ" file to be read from disk, only the json file I want to be passed to typst from memory.
> > 
> > 
> > yeah, you have to modify this line:
> > https://github.com/typst/typst/blob/b6a241cad232d920154986781c6f09d83dc1b6f6/cli/src/main.rs#L427
> 
> Maybe this solution works, if I can access the typst_library global module and create a new definition.
> 
> Then I would need to import into the typst/cli/main.rs some function from typst-library to convert my desired data into a typst value.
> 
> It is a bit hacky though.
> 
> Is this what you're thinking?
> 
> I will take a look at it.



> > Is this what you're thinking?
> 
> Mostly, except that you probably don't need to modify the Rust library's code, just take the output of the `build()` function, which will be a struct with a `.global` field which you can modify and add any variable you want there (it contains a `Scope` where you can use `.define("name", "something".into())`).
> 
> To be more specific, you can use `.global.scope_mut()` to get a mutable reference to the scope.

Yeah, that what was I thinking, however, to load a json it isn't as simple as "define("name", "something".into())."

The only change I did in the typst_library  was to make the typst_library::compute::convert_json function public.

What I did for now is at typst/cli/src/main.rs is:
```rust
        let mut library = typst_library::build();

        let my_value: Value = json!({
            "name": "John",
            "age": 30,
            "is_student": true,
            "hobbies": ["reading", "coding", "hiking"],
            "address": {
                "street": "123 Main St",
                "city": "Anytown",
                "state": "CA",
                "zip": "12345"
            }
        });

        library.global.scope_mut().define("data", typst_library::compute::convert_json(my_value));

        Self {
            root,
            library: Prehashed::new(library),
            book: Prehashed::new(searcher.book),
            fonts: searcher.fonts,
            hashes: RefCell::default(),
            paths: RefCell::default(),
            sources: FrozenVec::new(),
            main: SourceId::detached(),
        }

```

It is working!

Now I should bundle this as a library call.

Thanks!!

**Comment by Myriad-Dreamin** [comment]
I means you can intercept this function. When it reads a specific path, such as "/$mem", it will return the json content.
